data-science-onramp/data-cleaning/clean.py

```diff
@@ -0,0 +1,179 @@
+            except ValueError:
+                return None
+
+def compute_time(duration, start, end):
+    '''Calculates duration, start time, and end time from each other if one value is null.'''
+    time_format = '%Y-%m-%dT%H:%M:%S'
+
+    # Convert to datetime/timedelta
+    if start:
+        if '.' in start:
+            start = start[:start.index('.')]
+        start = datetime.datetime.strptime(start, time_format)
+    if end:
+        if '.' in end:
+            end = end[:end.index('.')]
+        end = datetime.datetime.strptime(end, time_format)
+    if duration:
+        duration = datetime.timedelta(seconds=duration)
+
+    # Compute null value
+    if start and end and not duration:
+        duration = end - start
+    elif duration and end and not start:
+        start = end - duration
+    elif duration and start and not end:
+        end = start + duration
+
+    # Convert from datetime/timedelta
+    if duration:
+        duration = int(duration.total_seconds())
+    if start:
+        start = start.strftime(time_format)
+    if end:
+        end = end.strftime(time_format)
+
+    return (duration, start, end)
+        
+def compute_duration(duration, start, end):
+    '''Calculates duration from start and end time if null.'''
+    return compute_time(duration, start, end)[0]
+
+def compute_start(duration, start, end):
+    '''Calculates start time from duration and end time if null.'''
+    return compute_time(duration, start, end)[1]
+
+def compute_end(duration, start, end):
+    '''Calculates end time from duration and start time if null.'''
+    return compute_time(duration, start, end)[2]
+
+def backfill(df):
+    path = 'gs://' + BUCKET_NAME + '/clean_citibike_data' + '.csv.gz'
+    df.write.options(codec='org.apache.hadoop.io.compress.GzipCodec').csv(path)
+
+def main():
+    # Create a SparkSession under the name 'clean'. Viewable via the Spark UI
+    spark = SparkSession.builder.appName('clean').getOrCreate()
+
+    # Whether to backfill data to GCS bucket
+    upload = True
+
+    # Check if table exists
+    try:
+        df = spark.read.format('bigquery').option('table', TABLE).load()
+    except Py4JJavaError:
+        print(f'{TABLE} does not exist. ')
+        return
+
+    # Single-parameter column transformations
+    udfs = {
+        'start_station_name': (station_name, StringType()),
+        'end_station_name': (station_name, StringType()),
+        'tripduration': (trip_duration, IntegerType()),
+        'usertype': (user_type, StringType()),
+        'gender': (gender, StringType()),
+        'start_station_latitude': (angle, FloatType()),
+        'start_station_longitude': (angle, FloatType()),
+        'end_station_latitude': (angle, FloatType()),
+        'end_station_longitude': (angle, FloatType())
+    }
+
+    for name, udf in udfs.items():
+        df = df.withColumn(name, UserDefinedFunction(*udf)(name))
+
+    # Multiple-parameter column transformations
+    multi_udfs = {
```

Comment:
Another multiple-parameter function Diego and I discussed when thinking of ways to dirty the script was for station ID and location/name. For example, if a station location or name is null, we were thinking of finding another row with the same station ID and getting the correct information from there if it's not null. There should not be any null station ID's because we did not touch that column